API not ideal when exclusively unsetting environment variables

I noticed that the API is a bit clumsy, when only unsetting variables.

The following does not compile:
```rust
#[test]
    fn test_function_works_when_env_var_missing() {
        temp_env::with_var("SHOULD_BE_MISSING", None, || {
            let actual = env::var("SHOULD_BE_MISSING");
            let expected = Result::Err(VarError::NotPresent);
            assert_eq!(expected, actual);
        })
    }
```

It only works like so:
```rust
#[test]
    fn test_function_works_when_env_var_missing() {
        //                vvvvvvvvvvvv
        temp_env::with_var::<_,&str,_>("SHOULD_BE_MISSING", None, || {
            let actual = env::var("SHOULD_BE_MISSING");
            let expected = Result::Err(VarError::NotPresent);
            assert_eq!(expected, actual);
        })
    }
```

The compiler error is not that nice either:
```
error[E0283]: type annotations needed
   --> src/test.rs:111:9
    |
111 |         temp_env::with_var("SHOULD_BE_MISSING", None, || {
    |         ^^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `V` declared on the function `with_var`
    |
    = note: cannot satisfy `_: AsRef<OsStr>`
note: required by a bound in `with_var`
   --> /Users/fabian.braun/.cargo/registry/src/github.com-1ecc6299db9ec823/temp-env-0.1.0/src/lib.rs:57:8
    |
57  |     V: AsRef<OsStr> + Clone,
    |        ^^^^^^^^^^^^ required by this bound in `with_var`
help: consider specifying the type arguments in the function call
    |
111 |         temp_env::with_var::<K, V, F>("SHOULD_BE_MISSING", None, || {
    |                           +++++++++++
```

This is a side-effect of using `AsRef<OsStr>` as a type bound instead of a concrete type. I assume you made that change to comply with the API of `env::with_var`. Do you have an idea how to go about this?

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -149,7 +149,7 @@ where
     let kvs = keys
         .iter()
         .map(|key| (key, None::<&str>))
-        .collect::<Vec<_>>();
+        .collect();
     with_vars(kvs, closure);
 }
 

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -146,10 +146,7 @@ where
     K: AsRef<OsStr> + Clone + Eq + Hash,
     F: Fn() + UnwindSafe + RefUnwindSafe,
 {
-    let kvs = keys
-        .iter()
-        .map(|key| (key, None::<&str>))
-        .collect();
+    let kvs = keys.iter().map(|key| (key, None::<&str>)).collect();
     with_vars(kvs, closure);
 }
 

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -296,8 +296,7 @@ mod tests {
         crate::with_vars_unset(vec!["SET_TO_BE_UNSET", "UNSET_TO_BE_UNSET"], || {
             assert_eq!(env::var("SET_TO_BE_UNSET"), Err(VarError::NotPresent));
             assert_eq!(env::var("UNSET_TO_BE_UNSET"), Err(VarError::NotPresent));
-        },
-        );
+        });
         assert_eq!(env::var("SET_TO_BE_UNSET"), Ok("val".to_string()));
         assert_eq!(env::var("UNSET_TO_BE_UNSET"), Err(VarError::NotPresent));
     }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -60,6 +60,25 @@ where
     with_vars(vec![(key, value)], closure)
 }
 
+/// Unsets a single environment variable for the duration of the closure.
+///
+/// The previous values are restored when the closure completes or panics, before unwinding the
+/// panic.
+///
+/// This is a shorthand and identical to the following:
+/// ```rust
+/// temp_env::with_var("MY_ENV_VAR", None::<&str>, || {
+///     // Run some code where `MY_ENV_VAR` is unset.
+/// });
+/// ```
+pub fn with_var_unset<K, F>(key: K, closure: F)
+where
+    K: AsRef<OsStr> + Clone + Eq + Hash,
+    F: Fn() + UnwindSafe + RefUnwindSafe,
+{
+    with_var(key, None::<&str>, closure);
+}
+
 /// Sets environment variables for the duration of the closure.
 ///
 /// The previous values are restored when the closure completes or panics, before unwinding the
@@ -143,7 +162,7 @@ mod tests {
         let foo_is_set = env::var("FOO").unwrap();
         assert_eq!(foo_is_set, "bar", "`FOO` must be set to \"bar\".");
 
-        crate::with_var("FOO", None::<&str>, || {
+        crate::with_var_unset("FOO", || {
             let foo_not_set = env::var("FOO");
             assert!(foo_not_set.is_err(), "`FOO` must not be set.");
         });

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -146,7 +146,10 @@ where
     K: AsRef<OsStr> + Clone + Eq + Hash,
     F: Fn() + UnwindSafe + RefUnwindSafe,
 {
-    let kvs = keys.iter().map(|key| (key, None::<&str>)).collect::<Vec<_>>();
+    let kvs = keys
+        .iter()
+        .map(|key| (key, None::<&str>))
+        .collect::<Vec<_>>();
     with_vars(kvs, closure);
 }
 
@@ -165,8 +168,8 @@ where
 // run in parallel.
 #[cfg(test)]
 mod tests {
-    use std::{env, panic};
     use std::env::VarError;
+    use std::{env, panic};
 
     /// Test whether setting a variable is correctly undone.
     #[test]
@@ -287,34 +290,16 @@ mod tests {
         env::set_var("SET_TO_BE_UNSET", "val");
         env::remove_var("UNSET_TO_BE_UNSET");
         // Check test preconditions
-        assert_eq!(
-            env::var("SET_TO_BE_UNSET"), Ok("val".to_string()),
-        );
-        assert_eq!(
-            env::var("UNSET_TO_BE_UNSET"),
-            Err(VarError::NotPresent),
-        );
+        assert_eq!(env::var("SET_TO_BE_UNSET"), Ok("val".to_string()));
+        assert_eq!(env::var("UNSET_TO_BE_UNSET"), Err(VarError::NotPresent));
 
-        crate::with_vars_unset(
-            vec!["SET_TO_BE_UNSET", "UNSET_TO_BE_UNSET"],
-            || {
-                assert_eq!(
-                    env::var("SET_TO_BE_UNSET"),
-                    Err(VarError::NotPresent),
-                );
-                assert_eq!(
-                    env::var("UNSET_TO_BE_UNSET"),
-                    Err(VarError::NotPresent),
-                );
-            },
+        crate::with_vars_unset(vec!["SET_TO_BE_UNSET", "UNSET_TO_BE_UNSET"], || {
+            assert_eq!(env::var("SET_TO_BE_UNSET"), Err(VarError::NotPresent));
+            assert_eq!(env::var("UNSET_TO_BE_UNSET"), Err(VarError::NotPresent));
+        },
         );
-        assert_eq!(
-            env::var("SET_TO_BE_UNSET"), Ok("val".to_string()),
-        );
-        assert_eq!(
-            env::var("UNSET_TO_BE_UNSET"),
-            Err(VarError::NotPresent),
-        );
+        assert_eq!(env::var("SET_TO_BE_UNSET"), Ok("val".to_string()));
+        assert_eq!(env::var("UNSET_TO_BE_UNSET"), Err(VarError::NotPresent));
     }
 
     /// Test whether unsetting one of the variable is correctly undone.

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -272,11 +272,11 @@ mod tests {
         env::remove_var("UNSET_TO_BE_UNSET");
         // Check test preconditions
         assert_eq!(
-            env::var("SET_TO_BE_UNSET"), Result::Ok("val".to_string()),
+            env::var("SET_TO_BE_UNSET"), Ok("val".to_string()),
         );
         assert_eq!(
             env::var("UNSET_TO_BE_UNSET"),
-            Result::Err(VarError::NotPresent),
+            Err(VarError::NotPresent),
         );
 
         crate::with_vars_unset(
@@ -284,20 +284,20 @@ mod tests {
             || {
                 assert_eq!(
                     env::var("SET_TO_BE_UNSET"),
-                    Result::Err(VarError::NotPresent),
+                    Err(VarError::NotPresent),
                 );
                 assert_eq!(
                     env::var("UNSET_TO_BE_UNSET"),
-                    Result::Err(VarError::NotPresent),
+                    Err(VarError::NotPresent),
                 );
             },
         );
         assert_eq!(
-            env::var("SET_TO_BE_UNSET"), Result::Ok("val".to_string()),
+            env::var("SET_TO_BE_UNSET"), Ok("val".to_string()),
         );
         assert_eq!(
             env::var("UNSET_TO_BE_UNSET"),
-            Result::Err(VarError::NotPresent),
+            Err(VarError::NotPresent),
         );
     }
 

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -185,18 +185,34 @@ mod tests {
 
     /// Test whether unsetting a variable is correctly undone.
     #[test]
-    fn test_with_var_unset() {
+    fn test_with_var_set_to_none() {
         env::set_var("FOO", "bar");
         let foo_is_set = env::var("FOO").unwrap();
         assert_eq!(foo_is_set, "bar", "`FOO` must be set to \"bar\".");
 
-        crate::with_var_unset("FOO", || {
+        crate::with_var("FOO", None::<&str>, || {
             let foo_not_set = env::var("FOO");
             assert!(foo_not_set.is_err(), "`FOO` must not be set.");
         });
 
         let foo_is_set_after = env::var("FOO").unwrap();
         assert_eq!(foo_is_set_after, "bar", "`FOO` must be set to \"bar\".");
+    }
+
+    /// Test whether unsetting a variable through the shorthand is correctly undone.
+    #[test]
+    fn test_with_var_unset() {
+        env::set_var("BAR", "foo");
+        let foo_is_set = env::var("BAR").unwrap();
+        assert_eq!(foo_is_set, "foo", "`BAR` must be set to \"foo\".");
+
+        crate::with_var_unset("BAR", || {
+            let foo_not_set = env::var("BAR");
+            assert!(foo_not_set.is_err(), "`BAR` must not be set.");
+        });
+
+        let foo_is_set_after = env::var("BAR").unwrap();
+        assert_eq!(foo_is_set_after, "foo", "`BAR` must be set to \"foo\".");
     }
 
     /// Test whether overriding an existing variable is correctly undone.

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -123,6 +123,33 @@ where
     };
 }
 
+/// Unsets environment variables for the duration of the closure.
+///
+/// The previous values are restored when the closure completes or panics, before unwinding the
+/// panic.
+///
+/// This is a shorthand and identical to the following:
+/// ```rust
+/// temp_env::with_vars(
+///     vec![
+///         ("FIRST_VAR", None::<&str>),
+///         ("SECOND_VAR", None::<&str>),
+///     ],
+///     || {
+///         // Run some code where `FIRST_VAR` and `SECOND_VAR` are unset (even if
+///         // they were set before)
+///     }
+/// );
+/// ```
+pub fn with_vars_unset<K, F>(keys: Vec<K>, closure: F)
+where
+    K: AsRef<OsStr> + Clone + Eq + Hash,
+    F: Fn() + UnwindSafe + RefUnwindSafe,
+{
+    let kvs = keys.iter().map(|key| (key, None::<&str>)).collect::<Vec<_>>();
+    with_vars(kvs, closure);
+}
+
 fn update_env<K, V>(key: K, value: Option<V>)
 where
     K: AsRef<OsStr>,
@@ -139,6 +166,7 @@ where
 #[cfg(test)]
 mod tests {
     use std::{env, panic};
+    use std::env::VarError;
 
     /// Test whether setting a variable is correctly undone.
     #[test]
@@ -237,9 +265,45 @@ mod tests {
         assert!(two_not_set_after.is_err(), "`TWO` must not be set.");
     }
 
-    /// Test whether unsetting one of the variable is correctly undone.
+    /// Test whether unsetting multiple variables is correctly undone.
     #[test]
     fn test_with_vars_unset() {
+        env::set_var("SET_TO_BE_UNSET", "val");
+        env::remove_var("UNSET_TO_BE_UNSET");
+        // Check test preconditions
+        assert_eq!(
+            env::var("SET_TO_BE_UNSET"), Result::Ok("val".to_string()),
+        );
+        assert_eq!(
+            env::var("UNSET_TO_BE_UNSET"),
+            Result::Err(VarError::NotPresent),
+        );
+
+        crate::with_vars_unset(
+            vec!["SET_TO_BE_UNSET", "UNSET_TO_BE_UNSET"],
+            || {
+                assert_eq!(
+                    env::var("SET_TO_BE_UNSET"),
+                    Result::Err(VarError::NotPresent),
+                );
+                assert_eq!(
+                    env::var("UNSET_TO_BE_UNSET"),
+                    Result::Err(VarError::NotPresent),
+                );
+            },
+        );
+        assert_eq!(
+            env::var("SET_TO_BE_UNSET"), Result::Ok("val".to_string()),
+        );
+        assert_eq!(
+            env::var("UNSET_TO_BE_UNSET"),
+            Result::Err(VarError::NotPresent),
+        );
+    }
+
+    /// Test whether unsetting one of the variable is correctly undone.
+    #[test]
+    fn test_with_vars_partially_unset() {
         let to_be_set_not_set = env::var("TO_BE_SET");
         assert!(to_be_set_not_set.is_err(), "`TO_BE_SET` must not be set.");
         env::set_var("TO_BE_UNSET", "unset");

